packages/utils/src/metrics.ts

```diff
@@ -0,0 +1,80 @@
+export type LabelsGeneric = Record<string, string | number>;
+export type LabelKeys<Labels extends LabelsGeneric> = Extract<keyof Labels, string>;
+export type CollectFn<Labels extends LabelsGeneric> = (metric: Gauge<Labels>) => void;
+
+// Type `{}` is required here as alternative types do not work
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type NoLabels = {};
+
+export interface Gauge<Labels extends LabelsGeneric = NoLabels> {
+  startTimer(): NoLabels extends Labels ? () => number : (labels: Labels) => number;
+  startTimer(labels: NoLabels extends Labels ? undefined : Labels): (labels?: Labels) => number;
+
+  inc: NoLabels extends Labels ? (value?: number) => void : (labels: Labels, value?: number) => void;
+  dec: NoLabels extends Labels ? (value?: number) => void : (labels: Labels, value?: number) => void;
+  set: NoLabels extends Labels ? (value: number) => void : (labels: Labels, value: number) => void;
+
+  collect?(): void;
+}
+
+export interface GaugeExtra<Labels extends LabelsGeneric = NoLabels> extends Gauge<Labels> {
+  addCollect(collectFn: CollectFn<Labels>): void;
+
+  /**
+   * @deprecated Use `addCollect` instead
+   */
+  collect(): void;
+}
+
+export interface Histogram<Labels extends LabelsGeneric = NoLabels> {
+  startTimer(): NoLabels extends Labels ? () => number : (labels: Labels) => number;
+  startTimer(labels: NoLabels extends Labels ? undefined : Labels): (labels?: Labels) => number;
+
+  observe: NoLabels extends Labels ? (value: number) => void : (labels: Labels, value: number) => void;
+
+  reset(): void;
+}
+
+export interface AvgMinMax<Labels extends LabelsGeneric = NoLabels> {
+  addGetValuesFn(getValuesFn: () => number[]): void;
+
+  set: NoLabels extends Labels ? (values: number[]) => void : (labels: Labels, values: number[]) => void;
+}
+
+export interface Counter<Labels extends LabelsGeneric = NoLabels> {
+  inc: NoLabels extends Labels ? (value?: number) => void : (labels: Labels, value?: number) => void;
+}
+
+export type GaugeConfig<Labels extends LabelsGeneric> = {
+  name: string;
+  help: string;
+} & (NoLabels extends Labels ? {labelNames?: never} : {labelNames: [LabelKeys<Labels>, ...LabelKeys<Labels>[]]});
+
+export type HistogramConfig<Labels extends LabelsGeneric> = GaugeConfig<Labels> & {
+  buckets?: number[];
+};
+
+export type AvgMinMaxConfig<Labels extends LabelsGeneric> = GaugeConfig<Labels>;
+
+export type CounterConfig<Labels extends LabelsGeneric> = GaugeConfig<Labels>;
+
+export type StaticConfig<Labels extends LabelsGeneric> = {
+  name: GaugeConfig<Labels>["name"];
+  help: GaugeConfig<Labels>["help"];
+  value: Record<LabelKeys<Labels>, string>;
+};
+
+export interface MetricsRegister {
+  gauge<Labels extends LabelsGeneric>(config: GaugeConfig<Labels>): Gauge<Labels>;
+  histogram<Labels extends LabelsGeneric>(config: HistogramConfig<Labels>): Histogram<Labels>;
+  counter<Labels extends LabelsGeneric>(config: CounterConfig<Labels>): Counter<Labels>;
```

Comment:
Should we not set default value for generics here, else we have to specify type everywhere we use the register.

Reply:
You don't have to specify the type explicitly, it is inferred from the configuration

e.g. see histogram usage without labels
https://github.com/ChainSafe/lodestar/blob/3b97f363c4ea6c5506e3534547840696a8473677/packages/beacon-node/src/metrics/metrics/beacon.ts#L114

But I still added them to address your previous comment https://github.com/ChainSafe/lodestar/pull/6201#discussion_r1431169824, see commit https://github.com/ChainSafe/lodestar/pull/6201/commits/2a14a4b0e2f22d03138ebae8fbc62ce5f991a79e